go.mod

```diff
@@ -71,9 +71,9 @@ require (
 	github.com/sethvargo/go-password v0.2.0
 	github.com/shirou/gopsutil/v3 v3.22.2
 	github.com/skratchdot/open-golang v0.0.0-20200116055534-eef842397966
-	github.com/spf13/cast v1.3.1
+	github.com/spf13/cast v1.5.0
 	github.com/spf13/cobra v1.5.0
-	github.com/spf13/viper v1.8.1
+	github.com/spf13/viper v1.12.0
```

Comment:
This upgrade is causing the CI failures.

https://github.com/spf13/viper/releases/tag/v1.12.0:
```
This release makes YAML v3 and TOML v2 the default versions used for encoding.
```

Which breaks the following tests:
```
go test -count 1 -v -run TestConfigOsqueryAsync
=== RUN   TestConfigOsqueryAsync
=== RUN   TestConfigOsqueryAsync/default
=== RUN   TestConfigOsqueryAsync/yaml_set_enabled_true
=== RUN   TestConfigOsqueryAsync/yaml_set_enabled_yes
    config_test.go:328:
                Error Trace:    /Users/luk/fleetdm/git/fleet/server/config/config_test.go:328
                Error:          func (assert.PanicTestFunc)(0x12a4640) should not panic
                                        Panic value:    Unable to cast to bool for key osquery.enable_async_host_processing: strconv.ParseBool: parsing "yes": invalid syntax
                                        Panic stack:    goroutine 53 [running]:
                                runtime/debug.Stack()
                                        /Users/luk/go/src/runtime/debug/stack.go:24 +0x65
                                github.com/stretchr/testify/assert.didPanic.func1()
                                        /Users/luk/gopath/pkg/mod/github.com/stretchr/testify@v1.8.0/assert/assertions.go:1050 +0x6c
                                panic({0x12cb040, 0xc00028f300})
                                        /Users/luk/go/src/runtime/panic.go:884 +0x212
                                github.com/fleetdm/fleet/v4/server/config.configForKeyOrBool.func1({0xc00025dbd0, 0x3})
                                        /Users/luk/fleetdm/git/fleet/server/config/config.go:1087 +0x25a
                                github.com/fleetdm/fleet/v4/server/config.configForKeyOrBool({0x1332988, 0x24}, {0x132a3df, 0x15}, {0xc00025dbd0, 0x3}, 0xd6?)
                                        /Users/luk/fleetdm/git/fleet/server/config/config.go:1094 +0xbb
                                github.com/fleetdm/fleet/v4/server/config.OsqueryConfig.AsyncConfigForTask({0x18, {0x1324f98, 0x8}, 0x0, {0x13259bf, 0xa}, {0x13259bf, 0xa}, 0x34630b8a000, 0x34630b8a000, ...}, ...)
                                        /Users/luk/fleetdm/git/fleet/server/config/config.go:160 +0x7d
                                github.com/fleetdm/fleet/v4/server/config.Manager.LoadConfig({_, _, _})
                                        /Users/luk/fleetdm/git/fleet/server/config/config.go:897 +0x2b45
                                github.com/fleetdm/fleet/v4/server/config.TestConfigOsqueryAsync.func1.2()
                                        /Users/luk/fleetdm/git/fleet/server/config/config_test.go:329 +0x46
                                github.com/stretchr/testify/assert.didPanic(0xc000082ea0?)
                                        /Users/luk/gopath/pkg/mod/github.com/stretchr/testify@v1.8.0/assert/assertions.go:1055 +0x8c
                                github.com/stretchr/testify/assert.NotPanics({0x13af1e0, 0xc000082ea0}, 0xc00058daa0, {0x0, 0x0, 0x0})
                                        /Users/luk/gopath/pkg/mod/github.com/stretchr/testify@v1.8.0/assert/assertions.go:1126 +0x7b
                                github.com/stretchr/testify/require.NotPanics({0x13b0090, 0xc000082ea0}, 0x0?, {0x0, 0x0, 0x0})
                                        /Users/luk/gopath/pkg/mod/github.com/stretchr/testify@v1.8.0/require/require.go:1486 +0x8b
                                github.com/fleetdm/fleet/v4/server/config.TestConfigOsqueryAsync.func1(0xc000082ea0)
                                        /Users/luk/fleetdm/git/fleet/server/config/config_test.go:328 +0x3bd
                                testing.tRunner(0xc000082ea0, 0xc0002603d0)
                                        /Users/luk/go/src/testing/testing.go:1446 +0x10b
                                created by testing.(*T).Run
                                        /Users/luk/go/src/testing/testing.go:1493 +0x35f
                Test:           TestConfigOsqueryAsync/yaml_set_enabled_yes
=== RUN   TestConfigOsqueryAsync/yaml_set_enabled_on
    config_test.go:328:
                Error Trace:    /Users/luk/fleetdm/git/fleet/server/config/config_test.go:328
                Error:          func (assert.PanicTestFunc)(0x12a4640) should not panic
                                        Panic value:    Unable to cast to bool for key osquery.enable_async_host_processing: strconv.ParseBool: parsing "on": invalid syntax
                                        Panic stack:    goroutine 6 [running]:
                                runtime/debug.Stack()
                                        /Users/luk/go/src/runtime/debug/stack.go:24 +0x65
                                github.com/stretchr/testify/assert.didPanic.func1()
                                        /Users/luk/gopath/pkg/mod/github.com/stretchr/testify@v1.8.0/assert/assertions.go:1050 +0x6c
                                panic({0x12cb040, 0xc0002f1a90})
                                        /Users/luk/go/src/runtime/panic.go:884 +0x212
                                github.com/fleetdm/fleet/v4/server/config.configForKeyOrBool.func1({0xc00001bc40, 0x2})
                                        /Users/luk/fleetdm/git/fleet/server/config/config.go:1087 +0x25a
                                github.com/fleetdm/fleet/v4/server/config.configForKeyOrBool({0x1332988, 0x24}, {0x1327c4c, 0x10}, {0xc00001bc40, 0x2}, 0xd6?)
                                        /Users/luk/fleetdm/git/fleet/server/config/config.go:1094 +0xbb
                                github.com/fleetdm/fleet/v4/server/config.OsqueryConfig.AsyncConfigForTask({0x18, {0x1324f98, 0x8}, 0x0, {0x13259bf, 0xa}, {0x13259bf, 0xa}, 0x34630b8a000, 0x34630b8a000, ...}, ...)
                                        /Users/luk/fleetdm/git/fleet/server/config/config.go:160 +0x7d
                                github.com/fleetdm/fleet/v4/server/config.Manager.LoadConfig({_, _, _})
                                        /Users/luk/fleetdm/git/fleet/server/config/config.go:897 +0x2b45
                                github.com/fleetdm/fleet/v4/server/config.TestConfigOsqueryAsync.func1.2()
                                        /Users/luk/fleetdm/git/fleet/server/config/config_test.go:329 +0x46
                                github.com/stretchr/testify/assert.didPanic(0xc0000831e0?)
                                        /Users/luk/gopath/pkg/mod/github.com/stretchr/testify@v1.8.0/assert/assertions.go:1055 +0x8c
                                github.com/stretchr/testify/assert.NotPanics({0x13af1e0, 0xc0000831e0}, 0xc00058ca20, {0x0, 0x0, 0x0})
                                        /Users/luk/gopath/pkg/mod/github.com/stretchr/testify@v1.8.0/assert/assertions.go:1126 +0x7b
                                github.com/stretchr/testify/require.NotPanics({0x13b0090, 0xc0000831e0}, 0x0?, {0x0, 0x0, 0x0})
                                        /Users/luk/gopath/pkg/mod/github.com/stretchr/testify@v1.8.0/require/require.go:1486 +0x8b
                                github.com/fleetdm/fleet/v4/server/config.TestConfigOsqueryAsync.func1(0xc0000831e0)
                                        /Users/luk/fleetdm/git/fleet/server/config/config_test.go:328 +0x3bd
                                testing.tRunner(0xc0000831e0, 0xc00028f3a0)
                                        /Users/luk/go/src/testing/testing.go:1446 +0x10b
                                created by testing.(*T).Run
                                        /Users/luk/go/src/testing/testing.go:1493 +0x35f
                Test:           TestConfigOsqueryAsync/yaml_set_enabled_on
=== RUN   TestConfigOsqueryAsync/yaml_set_enabled_invalid
=== RUN   TestConfigOsqueryAsync/yaml_set_enabled_per-task
=== RUN   TestConfigOsqueryAsync/yaml_set_invalid_per-task
=== RUN   TestConfigOsqueryAsync/envvar_set_enabled
=== RUN   TestConfigOsqueryAsync/envvar_set_enabled_on
=== RUN   TestConfigOsqueryAsync/envvar_set_enabled_per_task
=== RUN   TestConfigOsqueryAsync/yaml_collect_interval_lock_timeout
=== RUN   TestConfigOsqueryAsync/yaml_collect_interval_lock_timeout_per_task
=== RUN   TestConfigOsqueryAsync/yaml_env_var_override
--- FAIL: TestConfigOsqueryAsync (0.37s)
    --- PASS: TestConfigOsqueryAsync/default (0.04s)
    --- PASS: TestConfigOsqueryAsync/yaml_set_enabled_true (0.03s)
    --- FAIL: TestConfigOsqueryAsync/yaml_set_enabled_yes (0.04s)
    --- FAIL: TestConfigOsqueryAsync/yaml_set_enabled_on (0.04s)
    --- PASS: TestConfigOsqueryAsync/yaml_set_enabled_invalid (0.03s)
    --- PASS: TestConfigOsqueryAsync/yaml_set_enabled_per-task (0.02s)
    --- PASS: TestConfigOsqueryAsync/yaml_set_invalid_per-task (0.02s)
    --- PASS: TestConfigOsqueryAsync/envvar_set_enabled (0.02s)
    --- PASS: TestConfigOsqueryAsync/envvar_set_enabled_on (0.02s)
    --- PASS: TestConfigOsqueryAsync/envvar_set_enabled_per_task (0.03s)
    --- PASS: TestConfigOsqueryAsync/yaml_collect_interval_lock_timeout (0.02s)
    --- PASS: TestConfigOsqueryAsync/yaml_collect_interval_lock_timeout_per_task (0.03s)
    --- PASS: TestConfigOsqueryAsync/yaml_env_var_override (0.03s)
FAIL
exit status 1
FAIL    github.com/fleetdm/fleet/v4/server/config       0.790s
```

The fix as stated in the release is to build with the `viper_yaml2` tag:
```
go test -tags viper_yaml2 -count 1 -v -run TestConfigOsqueryAsync
=== RUN   TestConfigOsqueryAsync
=== RUN   TestConfigOsqueryAsync/default
=== RUN   TestConfigOsqueryAsync/yaml_set_enabled_true
=== RUN   TestConfigOsqueryAsync/yaml_set_enabled_yes
=== RUN   TestConfigOsqueryAsync/yaml_set_enabled_on
=== RUN   TestConfigOsqueryAsync/yaml_set_enabled_invalid
=== RUN   TestConfigOsqueryAsync/yaml_set_enabled_per-task
=== RUN   TestConfigOsqueryAsync/yaml_set_invalid_per-task
=== RUN   TestConfigOsqueryAsync/envvar_set_enabled
=== RUN   TestConfigOsqueryAsync/envvar_set_enabled_on
=== RUN   TestConfigOsqueryAsync/envvar_set_enabled_per_task
=== RUN   TestConfigOsqueryAsync/yaml_collect_interval_lock_timeout
=== RUN   TestConfigOsqueryAsync/yaml_collect_interval_lock_timeout_per_task
=== RUN   TestConfigOsqueryAsync/yaml_env_var_override
--- PASS: TestConfigOsqueryAsync (0.43s)
    --- PASS: TestConfigOsqueryAsync/default (0.03s)
    --- PASS: TestConfigOsqueryAsync/yaml_set_enabled_true (0.02s)
    --- PASS: TestConfigOsqueryAsync/yaml_set_enabled_yes (0.03s)
    --- PASS: TestConfigOsqueryAsync/yaml_set_enabled_on (0.03s)
    --- PASS: TestConfigOsqueryAsync/yaml_set_enabled_invalid (0.04s)
    --- PASS: TestConfigOsqueryAsync/yaml_set_enabled_per-task (0.07s)
    --- PASS: TestConfigOsqueryAsync/yaml_set_invalid_per-task (0.03s)
    --- PASS: TestConfigOsqueryAsync/envvar_set_enabled (0.02s)
    --- PASS: TestConfigOsqueryAsync/envvar_set_enabled_on (0.03s)
    --- PASS: TestConfigOsqueryAsync/envvar_set_enabled_per_task (0.02s)
    --- PASS: TestConfigOsqueryAsync/yaml_collect_interval_lock_timeout (0.03s)
    --- PASS: TestConfigOsqueryAsync/yaml_collect_interval_lock_timeout_per_task (0.03s)
    --- PASS: TestConfigOsqueryAsync/yaml_env_var_override (0.06s)
PASS
ok      github.com/fleetdm/fleet/v4/server/config       1.565s
```

Thus if we want to upgrade this library we should always build fleet with `viper_yaml2` build tag set.

Reply:
One solution is to use the build tag everywhere: [Add viper_yaml2 build tag everywhere](https://github.com/fleetdm/fleet/pull/7499/commits/9bbf4a4b61b95abd20d47f8baa4a0e6abc46ebd6).

Another solution is to stick to yaml v3, but that may be breaking change for users.

@chiiph @mna

Reply:
Well, that's annoying. Can't find anything yet that explains what the differences are between the two versions... and according to https://yaml.org/, the latest YAML spec is 1.2?